### ISSUE TITLE
Problem with OSX compilation

### ISSUE THREAD

**Issue by babaksamareh** [CoolProp/CoolProp#1248]
Guys,

I'm trying to compile the latest version of CoolProp in xCode to generate a static library for iPhone/iPad. I can compile version 5 with no problem, however, version 6 does not compile. I think some of the headers are called several times, and I get errors like:

/Users/babaksamareh/Desktop/CoolProp/CoolProp_v6.0.0_xCode/CoolProp/externals/rapidjson/include/rapidjson/msinttypes/stdint.h:38:2: "Use this header only with Microsoft Visual C++ compilers!"

or

/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator10.0.sdk/usr/include/sys/resource.h:195:2: Unknown type name 'uint8_t'

and also problems with ctime.h. Have you by any chance changed the way headers are called in the new version? Could it be the fact that OSX is case-insensitive and runs into problems?

Cheers,
Babak

PS: Not sure but it might be the cppformat in externals.

**Comment by ibell** [comment]
Weird, I use OSX all the time for development.  Is there something funny in
your workflow?  Are you using the development checkout of CoolProp?

We have bumped the revision for rapidjson, but our testing is looking good:
http://www.coolprop.dreamhosters.com:8010/grid (known failures mostly)

**Comment by ibell** [comment]
My guess is that the int types are getting redefined in Xcode (because of
iPhone architecture), and that is causing confusion for Xcode and integer
type clashes

**Comment by babaksamareh** [comment]
Well all the previous versions were working just fine. I am using the same instruction we had before, also posted on coolprop.org as iOS compilation (http://www.coolprop.org/coolprop/wrappers/iOS/index.html#ios). But that doesn't work anymore. Can you suggest any workaround?

**Comment by ibell** [comment]
What is the last version that works, and the newest version that causes
failure?  Does 6.0 work?

**Comment by ibell** [comment]
You could also try using the NO_CPPFORMAT preprocessor macro which will disable cppformat and use our old code for the format function

**Comment by babaksamareh** [comment]
5.1.2 works fine, but 6.0.0 does not. I'm sure it's the cppformat. Is there anyway to completely remove it from the code?

**Comment by babaksamareh** [comment]
And how do I use the NO_CPPFORMAT?

**Comment by ibell** [comment]
You would need to add an add_definitions(-DNO_CPPFORMAT) to the
CMakeLists.txt somewhere in the preamble.

**Comment by ibell** [comment]
Did that work?  We can also add a cmake flag so that you can set it directly yourself.

**Comment by babaksamareh** [comment]
ok I just managed to compile it, but I had to do lots of surgery. Here is what I did, first of all, culprits are cppformat and rapidjson. Apparently in xCode their headers mask system headers and cause confusion. I removed cppformat from externals and removed all the entries in the code. Next, I moved rapidjson from externals and added it to the include folder, as before. I also had to remove msinttypes in rapidjson. And now it works perfect. The problem was just header double definitions. That would be sweet if we can do something about this.

**Comment by babaksamareh** [comment]
Oh btw, in IF97.h (line 2033), xCode produces an error because one of the functions with a switch inside might not return a value! So I had to add return subregion; to the end of function.

**Comment by ibell** [comment]
Unfortunately, your problem only occurs in Xcode for iOS targets.  Can you
provide me with a simple way of debugging this issue so I can figure out
how to fix it properly?

Also, IF97 issue should be a warning, not an error, but I should fix that
anyway...

**Comment by babaksamareh** [comment]
I'm not sure what you mean by a simple way of debugging. I believe you use OSX, so you should have xCode installed, right? I can prepare the xCode project with all the errors and send it to you. Does this work?

**Comment by ibell** [comment]
Try that, but xcode native compilation works just fine, that's how I
develop CoolProp right now.

**Comment by babaksamareh** [comment]
I prepare the iOS compilation for you, you don't need an i-device to compile it, just compile it in the simulator mode, that should generate the 32/64 bit fat library that does not work on the actual devices but works fine in the simulator environment.

**Comment by babaksamareh** [comment]
Please also check this error, this only happens when compiling on actual device:
/externals/Eigen/Eigen/src/Core/arch/NEON/Complex.h:286 
Error: statement expression not allowed at file scope

I had to replace the line with this line and now it works:
static uint64x2_t p2ul_CONJ_XOR;// = vld1q_u64( p2ul_conj_XOR_DATA );

**Comment by babaksamareh** [comment]
[coolPropLib.zip](https://github.com/CoolProp/CoolProp/files/511528/coolPropLib.zip)

**Comment by ibell** [comment]
@babaksamareh - you should file the NEON bug back to Eigen people.  They are usually pretty responsive.

**Comment by ibell** [comment]
See also: http://stackoverflow.com/a/10713341/1360263 .  This might help to
keep everything in the cmake system.  I'll give it a try.

Can you please update the iOS instructions?  The "Targeted Device Family"
no longer seems to be a part of the configuration.

**Comment by ibell** [comment]
See my above commit, you would build the static library with:

```
cmake ../.. -G Xcode -DCOOLPROP_STATIC_LIBRARY=ON -DCOOLPROP_IOS_TARGET=ON

```
FWIW, it seems to build fine for me for the generic iOS target.

**Comment by babaksamareh** [comment]
Well if we can compile everything in cmake then we don't need the iOS tutorial anymore, I guess. Anyhow, I pulled the latest version, made a build directory, and inside build tried this:

cmake .. -G Xcode -DCOOLPROP_STATIC_LIBRARY=ON -DCOOLPROP_IOS_TARGET=ON

and this is the output:

-- COOLPROP_INSTALL_PREFIX=/Users/babaksamareh/Desktop/test/CoolProp/install_root
-- The C compiler identification is AppleClang 8.0.0.8000038
-- The CXX compiler identification is AppleClang 8.0.0.8000038
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- CoolProp version: 6.0.1dev
-- Found PythonInterp: /usr/bin/python2.7 (found suitable version "2.7.10", minimum required is "2.7") 
CMake Warning at CMakeLists.txt:158 (find_package):
  By not providing "FindREQUIRED.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "REQUIRED",
  but CMake did not find one.

  Could not find a package configuration file provided by "REQUIRED" with any
  of the following names:

    REQUIREDConfig.cmake
    required-config.cmake

  Add the installation prefix of "REQUIRED" to CMAKE_PREFIX_PATH or set
  "REQUIRED_DIR" to a directory containing one of the above files.  If
  "REQUIRED" provides a separate development package or SDK, be sure it has
  been installed.


-- Setting bitness flag -m64
-- Library compilation detected:
-- Creating CoolProp, a 64-bit library with Debug settings
-- CMAKE_SYSTEM_NAME: Darwin
-- COOLPROP_STATIC_LIBRARY: ON
-- COOLPROP_SHARED_LIBRARY: OFF
-- COOLPROP_OBJECT_LIBRARY: OFF
-- CONVENTION: 
-- COOLPROP_LIBRARY_HEADER: include/CoolPropLib.h
-- COOLPROP_LIBRARY_SOURCE: src/CoolPropLib.cpp
-- COMPILE_FLAGS:  -m64
-- LINK_FLAGS:  -m64
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/babaksamareh/Desktop/test/CoolProp/build
mie-13-62:build babaksamareh$ 

However, when I tried to make it I get this error:

make: *** No targets specified and no makefile found.  Stop.

**Comment by ibell** [comment]
What I did:

Ians-Mac-mini:iOS ian$ cmake ../.. -G Xcode -DCOOLPROP_STATIC_LIBRARY=ON
-DCOOLPROP_IOS_TARGET=ON

-- COOLPROP_INSTALL_PREFIX=/Users/ian/Code/coolprop/install_root

-- The C compiler identification is AppleClang 7.0.2.7000181

-- The CXX compiler identification is AppleClang 7.0.2.7000181

-- Check for working C compiler using: Xcode

-- Check for working C compiler using: Xcode -- works

-- Detecting C compiler ABI info

-- Detecting C compiler ABI info - done

-- Detecting C compile features

-- Detecting C compile features - done

-- Check for working CXX compiler using: Xcode

-- Check for working CXX compiler using: Xcode -- works

-- Detecting CXX compiler ABI info

-- Detecting CXX compiler ABI info - done

-- Detecting CXX compile features

-- Detecting CXX compile features - done

-- CoolProp version: 6.0.1dev

-- Found PythonInterp: /Users/ian/anaconda/bin/python2.7 (found suitable
version "2.7.11", minimum required is "2.7")

CMake Warning at CMakeLists.txt:158 (find_package):

  By not providing "FindREQUIRED.cmake" in CMAKE_MODULE_PATH this project
has

  asked CMake to find a package configuration file provided by "REQUIRED",

  but CMake did not find one.


  Could not find a package configuration file provided by "REQUIRED" with
any

  of the following names:


    REQUIREDConfig.cmake

    required-config.cmake


  Add the installation prefix of "REQUIRED" to CMAKE_PREFIX_PATH or set

  "REQUIRED_DIR" to a directory containing one of the above files.  If

  "REQUIRED" provides a separate development package or SDK, be sure it has

  been installed.



-- Setting bitness flag -m64

-- Library compilation detected:

-- Creating CoolProp, a 64-bit library with Debug settings

-- CMAKE_SYSTEM_NAME: Darwin

-- COOLPROP_STATIC_LIBRARY: ON

-- COOLPROP_SHARED_LIBRARY: OFF

-- COOLPROP_OBJECT_LIBRARY: OFF

-- CONVENTION:

-- COOLPROP_LIBRARY_HEADER: include/CoolPropLib.h

-- COOLPROP_LIBRARY_SOURCE: src/CoolPropLib.cpp

-- COMPILE_FLAGS:  -m64

-- LINK_FLAGS:  -m64

-- Configuring done

-- Generating done

-- Build files have been written to: /Users/ian/Code/coolprop/build/iOS


In Xcode:

1. Switched build to CoolProp -> Generic iOS Device

2. Build!


It worked fine for me.  Handful of warnings, but no errors.

**Comment by babaksamareh** [comment]
Oohh interesting, it works :D GREAT
So, we need to update the instructions. Now I can update the tutorial for iOS, but I have never used github before, it's a shame I know, so before I mess something up, just quickly tell me what to do.

**Comment by babaksamareh** [comment]
Ok I guess I did it under my own fork, please find it here:
https://github.com/babaksamareh/CoolProp/blob/master/Web/coolprop/wrappers/iOS/index.rst

PS: For future reference, after I edit my own fork, how do I merge it with the main project?

**Comment by ibell** [comment]
To merge, you file a pull request between your fork and the main
repository.  There are good docs on github website about how to do this.
It is best if you make your changes in a branch, because it is easier to
then rebase from the main repo after the pull request has been accepted.

On Oct 7, 2016 8:50 AM, "Babak Samareh" <notifications@github.com> wrote:

Ok I guess I did it under my own fork, please find it here:
https://github.com/babaksamareh/CoolProp/blob/master/Web/coolprop/wrappers/
iOS/index.rst

PS: For future reference, after I edit my own fork, how do I merge it with
the main project?

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/CoolProp/CoolProp/issues/1248#issuecomment-252273043>,
or mute the thread

**Comment by babaksamareh** [comment]
Perfect I made the pull request. Not sure if I did it right, please excuse if there is any mistakes. I have already started working with github and trying to upload my projects :) 
Thanks again.

**Comment by ibell** [comment]
Looks good, I merged it!

**Comment by ibell** [comment]
Closed by #1257
